cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
-			$this->plugin->components['media']->get_post_meta( $post_id, self::META_KEYS['sync_error'], $upload->get_error_message() );
+	public function generate_signature( $attachment_id, $cache = true ) {
+		static $signatures = array(); // cache signatures.
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cache ) {
+			$return = $signatures[ $attachment_id ];
+		} else {
+			$return = $this->sync_base( $attachment_id );
+			if ( ! is_wp_error( $return ) ) {
+				$return                       = array_map(
+					function ( $item ) {
+						if ( is_array( $item ) ) {
+							$item = wp_json_encode( $item );
+						}
 
-			return false;
+						return md5( $item );
+					},
+					$return
+				);
+				$signatures[ $attachment_id ] = $return;
+			}
 		}
-		$credentials          = $this->plugin->components['connect']->get_credentials();
-		$upload['cloud_name'] = $credentials['cloud_name'];
-		$return               = array_map(
-			function ( $item ) {
-				if ( is_array( $item ) ) {
-					$item = wp_json_encode( $item );
-				}
-
-				return md5( $item );
-			},
-			$upload
-		);
 
 		return $return;
 	}
 
+	/**
+	 * Check if an asset can be synced.
+	 *
+	 * @param int    $attachment_id The attachment ID to check if it  can be synced.
+	 * @param string $type          The type of sync to attempt.
+	 *
+	 * @return bool
+	 */
+	public function can_sync( $attachment_id, $type = 'file' ) {
+
+		$can = $this->is_auto_sync_enabled();
+
+		if ( $this->is_pending( $attachment_id ) ) {
+			$can = false;
+		} elseif ( $this->been_synced( $attachment_id ) ) {
+			$can = true;
+		}
+
+		/**
+		 * Filter to allow changing if an asset is allowed to be synced.
+		 * Return a WP Error with reason why it can't be synced.
+		 *
+		 * @param int $attachment_id The attachment post ID.
+		 *
+		 * @return bool|\WP_Error
+		 */
+		return apply_filters( 'cloudinary_can_sync_asset', $can, $attachment_id, $type );
+	}
+
+	/**
+	 * Get the last version this asset was synced with.
+	 *
+	 * @param int $attachment_id The attachment ID.
+	 *
+	 * @return mixed
+	 */
+	public function get_sync_version( $attachment_id ) {
+		$version = $this->managers['media']->get_post_meta( $attachment_id, self::META_KEYS['plugin_version'], true );
+
+		return $version;
+	}
+
 	/**
 	 * Get the current sync signature of an asset.
 	 *
-	 * @param int $post_id The post ID.
+	 * @param int  $attachment_id The attachment ID.
+	 * @param bool $cached        Flag to specify if a cached signature is to be used or build a new one.
 	 *
 	 * @return array|bool
 	 */
-	public function get_signature( $post_id ) {
+	public function get_signature( $attachment_id, $cached = true ) {
 		static $signatures = array(); // Cache signatures already fetched.
 
-		$return = false;
-		if ( ! empty( $signatures[ $post_id ] ) ) {
-			$return = $signatures[ $post_id ];
+		$return = array();
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cached ) {
+			$return = $signatures[ $attachment_id ];
```

Comment:
Same as above, we can return here and avoid the `else` nesting, making the code more readable.

Reply:
Same as above. trying to maintain a single exit point to aid in code coverage.